pkg/doc/did/doc.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package did
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	jsonldType         = "type"
+	jsonldID           = "id"
+	jsonldServicePoint = "serviceEndpoint"
+	jsonldController   = "controller"
+
+	// various public key encodings
+	jsonldPublicKeyBase64 = "publicKeyBase64"
+	jsonldPublicKeyBase58 = "publicKeyBase58"
+	jsonldPublicKeyHex    = "publicKeyHex"
+	jsonldPublicKeyPem    = "publicKeyPem"
+	jsonldPublicKeyJwk    = "publicKeyJwk"
+)
+
+// Doc DID Document definition
+type Doc struct {
+	Context        []string      `json:"@context,omitempty"`
+	ID             string        `json:"id,omitempty"`
+	PublicKey      []PublicKey   `json:"publicKey,omitempty"`
+	Service        []Service     `json:"service,omitempty"`
+	Authentication []interface{} `json:"authentication,omitempty"`
+	Created        string        `json:"created,omitempty"`
+	Updated        string        `json:"updated,omitempty"`
+	Proof          Proof         `json:"proof,omitempty"`
+}
+
+// Proof is cryptographic proof of the integrity of the DID Document
+type Proof struct {
+	Type           string `json:"type,omitempty"`
+	Created        string `json:"created,omitempty"`
+	Creator        string `json:"creator,omitempty"`
+	SignatureValue string `json:"signatureValue,omitempty"`
+	Domain         string `json:"domain,omitempty"`
+	Nonce          string `json:"nonce,omitempty"`
+}
+
+// FromBytes creates an instance of DIDDocument by reading a JSON document from bytes
+func FromBytes(data []byte) (*Doc, error) {
+
+	doc := &Doc{}
+	err := json.Unmarshal(data, &doc)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal did doc bytes")
+	}
+	// validate did document
+	if err := doc.validate(); err != nil {
+		return nil, err
+	}
+	return doc, nil
+}
+
+// validate did doc
+func (doc *Doc) validate() error {
+	// TODO Validate that the output DID Document conforms to the serialization of the DID Document data model (https://w3c-ccg.github.io/did-spec/#did-documents)
+	return nil
+}
+
+// stringEntry
+func stringEntry(entry interface{}) string {
+	if entry == nil {
+		return ""
+	}
+	id, ok := entry.(string)
+	if !ok {
```

Comment:
when does this happen?

Reply:
This function is reused in several places to extract values for attributes of type string. I'm concerned about how this function assumes it's ok to return an empty string if the input is not a string. This function should just not be used if the input were not a string (alternatively, it could provide a boolean return parameter that would indicate as much)